File: backend/app/schemas/progress.py
from pydantic import BaseModel
from typing import Optional, List
from uuid import UUID
from datetime import datetime


class ProgressBase(BaseModel):
    plan_id: UUID
    group_id: Optional[UUID] = None
    week_number: int
    report_date: datetime
    talk_count: int = 0
    doc_review_count: int = 0
    petition_count: int = 0
    visit_count: int = 0
    problem_total: int = 0
    problem_party: int = 0
    problem_pty: int = 0
    problem_key: int = 0
    next_week_plan: Optional[str] = None
    notes: Optional[str] = None


class ProgressCreate(ProgressBase):
    pass


class ProgressUpdate(BaseModel):
    group_id: Optional[UUID] = None
    week_number: Optional[int] = None
    report_date: Optional[datetime] = None
    talk_count: Optional[int] = None
    doc_review_count: Optional[int] = None
    petition_count: Optional[int] = None
    visit_count: Optional[int] = None
    problem_total: Optional[int] = None
    problem_party: Optional[int] = None
    problem_pty: Optional[int] = None
    problem_key: Optional[int] = None
    next_week_plan: Optional[str] = None
    notes: Optional[str] = None


class ProgressResponse(ProgressBase):
    id: UUID
    is_active: bool
    created_by: UUID
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class GroupOverview(BaseModel):
    """Dashboard card overview by group"""
    group_id: Optional[UUID] = None
    group_name: Optional[str] = None
    plan_id: UUID
    plan_name: Optional[str] = None
    total_reports: int = 0
    total_talks: int = 0
    total_doc_reviews: int = 0
    total_petitions: int = 0
    total_visits: int = 0
    total_problems: int = 0
    latest_report_date: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/api/progress.ts
import api from './client';

export interface ProgressRecord {
  id: string;
  plan_id: string;
  plan_name?: string;
  group_id?: string;
  group_name?: string;
  week_number: number;
  report_date: string;
  talk_count: number;
  doc_review_count: number;
  petition_count: number;
  visit_count: number;
  problem_total: number;
  problem_party: number;
  problem_pty: number;
  problem_key: number;
  next_week_plan: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export const getProgressList = (params?: { plan_id?: string; group_id?: string; page?: number; page_size?: number }) =>
  api.get('/progress/', { params }).then(res => res.data);

// Backend returns {data: {progress_fields}, message: "..."} — manual unwrap needed.
export const getProgress = (id: string) =>
  api.get(`/progress/${id}`).then(res => (res.data as any).data);

export const createProgress = (data: any) =>
  api.post('/progress/', data).then(res => res.data);

export const updateProgress = (id: string, data: any) =>
  api.put(`/progress/${id}`, data).then(res => res.data);

export const deleteProgress = (id: string) =>
  api.delete(`/progress/${id}`);

export const importProgress = (planId: string, file: File) => {
  const formData = new FormData();
  formData.append('file', file);
  return api.post(`/progress/import/${planId}`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  }).then(res => res.data);
};

export const downloadProgressTemplate = (planId?: string) => {
  return api.get('/progress/template', { 
    params: planId ? { plan_id: planId } : {},
    responseType: 'blob' 
  }).then(res => {
    const blob = new Blob([res.data], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = '巡察进度导入模板.xlsx';
    a.click();
    window.URL.revokeObjectURL(url);
  });
};

export const exportProgress = (params?: { plan_id?: string; group_id?: string; year?: number }) => {
  return api.get('/progress/export', { params, responseType: 'blob' }).then(res => {
    const blob = new Blob([res.data], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = '巡察进度导出.xlsx';
    a.click();
    window.URL.revokeObjectURL(url);
  });
};
